src/repr/src/adt/interval.rs

```diff
@@ -272,16 +294,57 @@ impl Interval {
         (self.micros % 1_000_000 * 1_000).try_into().unwrap()
     }
 
-    /// Computes the total number of seconds in the interval.
-    pub fn as_seconds<T>(&self) -> T
+    /// Computes the total number of epoch seconds in the interval.
+    /// When extracting an epoch, PostgreSQL considers a year
+    /// 365.25 days.
+    pub fn as_epoch_seconds<T>(&self) -> T
     where
         T: DecimalLike,
     {
-        T::from(self.years() * 60 * 60 * 24) * T::from(365.25)
-            + T::from(self.months() * 60 * 60 * 24 * 30)
-            + T::from(self.days * 60 * 60 * 24)
-            + T::lossy_from(self.dur_as_secs())
-            + T::from(self.nanoseconds()) / T::from(1e9)
+        let days = T::from(self.years()) * T::from(Self::EPOCH_DAYS_PER_YEAR)
+            + T::from(self.months()) * T::from(Self::DAY_PER_MONTH)
+            + T::from(self.days);
+        let seconds =
+            days * T::from(Self::HOUR_PER_DAY * Self::MINUTE_PER_HOUR * Self::SECOND_PER_MINUTE);
+
+        seconds
+            + T::lossy_from(self.micros)
+                / T::from(Self::MICROSECOND_PER_MILLISECOND * Self::MILLISECOND_PER_SECOND)
+    }
+
+    /// Computes the total number of microseconds in the interval.
+    pub fn as_microseconds(&self) -> i128 {
+        // unwrap is safe because i32::MAX/i32::MIN number of months will not overflow an i128 when
+        // converted to microseconds.
+        Self::convert_date_time_unit(
+            DateTimeField::Month,
+            DateTimeField::Microseconds,
+            i128::from(self.months),
+        ).unwrap() +
+        // unwrap is safe because i32::MAX/i32::MIN number of days will not overflow an i128 when
+        // converted to microseconds.
+        Self::convert_date_time_unit(
+            DateTimeField::Day,
+            DateTimeField::Microseconds,
+            i128::from(self.days),
+        ).unwrap() +
+        i128::from(self.micros)
+    }
+
+    /// Converts this `Interval`'s duration into `chrono::Duration`.
+    pub fn duration_as_chrono(&self) -> chrono::Duration {
+        use chrono::Duration;
+        Duration::days(self.days.into()) + Duration::microseconds(self.micros)
+    }
+
+    pub fn duration(&self) -> Result<Duration, anyhow::Error> {
+        if self.months != 0 {
+            bail!("cannot convert interval with months to duration");
+        }
+        if self.duration_as_chrono() < chrono::Duration::zero() {
+            bail!("cannot convert negative interval to duration");
+        }
+        Ok(self.duration_as_chrono().to_std()?)
```

Comment:
What does chrono buy us here as we're immediately throwing it away? The negative detection?

Reply:
It was purely for convenience to try and avoid code duplication. I've removed it.